File: Data Structures & Algorithms/clone-graph/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    void dfs(unordered_map<Node*, Node*>& visited, Node* node){
        if(visited.count(node)){
            return;
        }

        Node* clone = new Node(node->val);
        visited[node] = clone;

        for(Node* nei : node->neighbors){
            if(!visited.count(nei)){
                dfs(visited, nei);
            }
            clone->neighbors.push_back(visited[nei]);
        }
    }
    Node* cloneGraph(Node* node) {
        if(!node) return NULL;
        unordered_map<Node*, Node*> visited;
        dfs(visited, node);
        return visited[node];
    }
};

File: Data Structures & Algorithms/count-connected-components/submission-0.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for n1, n2 in edges:
            graph[n2].append(n1)
            graph[n1].append(n2)

        def dfs(node):
            if node in visited:
                return
            
            visited.add(node)
            for n in graph[node]:
                dfs(n)

        count = 0
        visited = set()
        for i in range(n):
            if i not in visited:
                count += 1
                dfs(i)
        
        return count